File: Leetcode/Backtrace/39_combination-sum/main.cpp
//  39. 组合总和
// Created by bob on 2022/6/20.
//
#include<vector>

using namespace std;

class Solution {
public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<vector<int>> result;
        vector<int> current;
        dfs(candidates, target, 0, result, current);
        return result;
    }

    void dfs(vector<int>& nums, int target, int idx, vector<vector<int>> &result, vector<int> &cur) {
        if (idx == nums.size()) {
            return;
        }
        if (target == 0) {
            result.push_back(cur);
            return;
        }
        // 直接跳过
        dfs(nums, target, idx + 1, result, cur);
        // 选择当前数
        if (target - nums[idx] >= 0) {
            cur.push_back(nums[idx]);
            dfs(nums, target - nums[idx], idx, result, cur);
            cur.pop_back();
        }

    }
};

int main() {
    vector<int> candidates = {2, 3, 6, 7};
    int target = 7;
    vector<vector<int>> result = Solution().combinationSum(candidates, target);
    return 0;
}

File: Leetcode/Backtrace/39_combination-sum/main.py
from typing import List


class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        results = list()
        current = []

        def dfs(idx: int, tar: int):
            if idx == len(candidates):
                return
            if tar == 0:
                results.append(current[:])   # 注：results.append(current) 写法是错误的，相当于浅。但是需求是只需要数组里的值，而不是引用
                return
            dfs(idx + 1, tar)  # 直接跳过
            if tar - candidates[idx] >= 0:
                current.append(candidates[idx])
                tar -= candidates[idx]
                dfs(idx, tar)
                current.pop()
        dfs(0, target)
        return results

if __name__ == '__main__':
    candidates = [2, 3, 6, 7]
    target = 7
    results = Solution().combinationSum(candidates, target)
    print(results)
